DirectionsRoute#fromJson RuntimeException

Originally reported in https://github.com/mapbox/mapbox-navigation-android/issues/1199

> Run into the following 💥 while testing
>
> ```
> 08-08 13:57:37.398 11915-11915/com.mapbox.services.android.navigation.testapp E/NativeMapView: Exception in MapView.OnMapChangedListener
> java.lang.RuntimeException: Failed to invoke public com.mapbox.geojson.Point() with no args
> at com.google.gson.internal.ConstructorConstructor$3.construct(ConstructorConstructor.java:113)
> at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:212)
> at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
> at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
> at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
> at com.mapbox.api.directions.v5.models.AutoValue_RouteOptions$GsonTypeAdapter.read(AutoValue_RouteOptions.java:339)
> at com.mapbox.api.directions.v5.models.AutoValue_RouteOptions$GsonTypeAdapter.read(AutoValue_RouteOptions.java:35)
> at com.mapbox.api.directions.v5.models.AutoValue_DirectionsRoute$GsonTypeAdapter.read(AutoValue_DirectionsRoute.java:205)
> at com.mapbox.api.directions.v5.models.AutoValue_DirectionsRoute$GsonTypeAdapter.read(AutoValue_DirectionsRoute.java:30)
> at com.google.gson.Gson.fromJson(Gson.java:927)
> at com.google.gson.Gson.fromJson(Gson.java:892)
> at com.google.gson.Gson.fromJson(Gson.java:841)
> at com.google.gson.Gson.fromJson(Gson.java:813)
> at com.mapbox.api.directions.v5.models.DirectionsRoute.fromJson(DirectionsRoute.java:147)
> at com.mapbox.services.android.navigation.ui.v5.NavigationLauncher.extractRoute(NavigationLauncher.java:62)
> at com.mapbox.services.android.navigation.ui.v5.MapboxNavigationActivity.extractRoute(MapboxNavigationActivity.java:122)
> at com.mapbox.services.android.navigation.ui.v5.MapboxNavigationActivity.onNavigationReady(MapboxNavigationActivity.java:98)
> at com.mapbox.services.android.navigation.ui.v5.NavigationView.onMapReady(NavigationView.java:203)
> at com.mapbox.mapboxsdk.maps.MapView$MapCallback.onMapReady(MapView.java:1157)
> at com.mapbox.mapboxsdk.maps.MapView$MapCallback.onMapChanged(MapView.java:1142)
> at com.mapbox.mapboxsdk.maps.NativeMapView.onMapChanged(NativeMapView.java:900)
> at android.os.MessageQueue.nativePollOnce(Native Method)
> at android.os.MessageQueue.next(MessageQueue.java:323)
> at android.os.Looper.loop(Looper.java:135)
> at android.app.ActivityThread.main(ActivityThread.java:5417)
> at java.lang.reflect.Method.invoke(Native Method)
> at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
> Caused by: java.lang.InstantiationException: Can't instantiate abstract class com.mapbox.geojson.Point
> at java.lang.reflect.Constructor.newInstance(Native Method)
> at com.google.gson.internal.ConstructorConstructor$3.construct(ConstructorConstructor.java:110)
> at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:212) 
> at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41) 
> at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82) 
> at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61) 
> at com.mapbox.api.directions.v5.models.AutoValue_RouteOptions$GsonTypeAdapter.read(AutoValue_RouteOptions.java:339) 
> at com.mapbox.api.directions.v5.models.AutoValue_RouteOptions$GsonTypeAdapter.read(AutoValue_RouteOptions.java:35) 
> at com.mapbox.api.directions.v5.models.AutoValue_DirectionsRoute$GsonTypeAdapter.read(AutoValue_DirectionsRoute.java:205) 
> at com.mapbox.api.directions.v5.models.AutoValue_DirectionsRoute$GsonTypeAdapter.read(AutoValue_DirectionsRoute.java:30) 
> at com.google.gson.Gson.fromJson(Gson.java:927) 
> at com.google.gson.Gson.fromJson(Gson.java:892) 
> at com.google.gson.Gson.fromJson(Gson.java:841) 
> at com.google.gson.Gson.fromJson(Gson.java:813) 
> at com.mapbox.api.directions.v5.models.DirectionsRoute.fromJson(DirectionsRoute.java:147) 
> at com.mapbox.services.android.navigation.ui.v5.NavigationLauncher.extractRoute(NavigationLauncher.java:62) 
> at com.mapbox.services.android.navigation.ui.v5.MapboxNavigationActivity.extractRoute(MapboxNavigationActivity.java:122) 
> at com.mapbox.services.android.navigation.ui.v5.MapboxNavigationActivity.onNavigationReady(MapboxNavigationActivity.java:98) 
> at com.mapbox.services.android.navigation.ui.v5.NavigationView.onMapReady(NavigationView.java:203) 
> at com.mapbox.mapboxsdk.maps.MapView$MapCallback.onMapReady(MapView.java:1157) 
> at com.mapbox.mapboxsdk.maps.MapView$MapCallback.onMapChanged(MapView.java:1142) 
> at com.mapbox.mapboxsdk.maps.NativeMapView.onMapChanged(NativeMapView.java:900) 
> at android.os.MessageQueue.nativePollOnce(Native Method) 
> at android.os.MessageQueue.next(MessageQueue.java:323) 
> at android.os.Looper.loop(Looper.java:135) 
> at android.app.ActivityThread.main(ActivityThread.java:5417) 
> at java.lang.reflect.Method.invoke(Native Method) 
> at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
> ```
>
> Pretty easy to reproduce launching a navigation session using the `NavigationLauncher` 👉 [`NavigationLauncherActivity`](https://github.com/mapbox/mapbox-navigation-android/blob/master/app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/navigationui/NavigationLauncherActivity.java)
>
> I think introduced by the latest changes in MAS `3.4.0` https://github.com/mapbox/mapbox-navigation-android/pull/1196

> > It looks like the old `DirectionsRoute#fromJson` had some extra adapters / deserializers that were removed https://github.com/mapbox/mapbox-java/pull/854/files#diff-a19faf46d28dec2e35e529c01f8408bdL49
>
> Yeah that seems the issue.

cc @danesfeder

--- a/services-directions/src/main/java/com/mapbox/api/directions/v5/models/DirectionsJsonObject.java
+++ b/services-directions/src/main/java/com/mapbox/api/directions/v5/models/DirectionsJsonObject.java
@@ -2,6 +2,8 @@ package com.mapbox.api.directions.v5.models;
 
 import com.google.gson.GsonBuilder;
 import com.mapbox.api.directions.v5.DirectionsAdapterFactory;
+import com.mapbox.geojson.Point;
+import com.mapbox.geojson.gson.PointSerializer;
 
 import java.io.Serializable;
 
@@ -22,6 +24,7 @@ public class DirectionsJsonObject implements Serializable {
   public String toJson() {
     GsonBuilder gson = new GsonBuilder();
     gson.registerTypeAdapterFactory(DirectionsAdapterFactory.create());
+    gson.registerTypeAdapter(Point.class, new PointSerializer());
     return gson.create().toJson(this);
   }
 }

--- a/services-directions/src/main/java/com/mapbox/api/directions/v5/models/DirectionsResponse.java
+++ b/services-directions/src/main/java/com/mapbox/api/directions/v5/models/DirectionsResponse.java
@@ -7,6 +7,8 @@ import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.google.gson.TypeAdapter;
 import com.mapbox.api.directions.v5.DirectionsAdapterFactory;
+import com.mapbox.geojson.Point;
+import com.mapbox.geojson.gson.PointDeserializer;
 
 import java.util.List;
 
@@ -126,6 +128,7 @@ public abstract class DirectionsResponse extends DirectionsJsonObject {
   public static DirectionsResponse fromJson(String json) {
     GsonBuilder gson = new GsonBuilder();
     gson.registerTypeAdapterFactory(DirectionsAdapterFactory.create());
+    gson.registerTypeAdapter(Point.class, new PointDeserializer());
     return gson.create().fromJson(json, DirectionsResponse.class);
   }
 

--- a/services-directions/src/main/java/com/mapbox/api/directions/v5/models/DirectionsRoute.java
+++ b/services-directions/src/main/java/com/mapbox/api/directions/v5/models/DirectionsRoute.java
@@ -9,6 +9,8 @@ import com.google.gson.TypeAdapter;
 import com.google.gson.annotations.SerializedName;
 import com.mapbox.api.directions.v5.DirectionsAdapterFactory;
 import com.mapbox.api.directions.v5.MapboxDirections;
+import com.mapbox.geojson.Point;
+import com.mapbox.geojson.gson.PointDeserializer;
 
 import java.util.List;
 
@@ -144,6 +146,7 @@ public abstract class DirectionsRoute extends DirectionsJsonObject {
   public static DirectionsRoute fromJson(String json) {
     GsonBuilder gson = new GsonBuilder();
     gson.registerTypeAdapterFactory(DirectionsAdapterFactory.create());
+    gson.registerTypeAdapter(Point.class, new PointDeserializer());
     return gson.create().fromJson(json, DirectionsRoute.class);
   }
 

--- a/services-directions/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java
+++ b/services-directions/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java
@@ -11,6 +11,7 @@ import com.mapbox.api.directions.v5.DirectionsAdapterFactory;
 import com.mapbox.api.directions.v5.DirectionsCriteria;
 import com.mapbox.api.directions.v5.MapboxDirections;
 import com.mapbox.geojson.Point;
+import com.mapbox.geojson.gson.PointDeserializer;
 
 import java.util.List;
 
@@ -289,10 +290,11 @@ public abstract class RouteOptions extends DirectionsJsonObject {
    *   method
    * @since 3.4.0
    */
-  public static RouteLeg fromJson(String json) {
+  public static RouteOptions fromJson(String json) {
     GsonBuilder gson = new GsonBuilder();
     gson.registerTypeAdapterFactory(DirectionsAdapterFactory.create());
-    return gson.create().fromJson(json, RouteLeg.class);
+    gson.registerTypeAdapter(Point.class, new PointDeserializer());
+    return gson.create().fromJson(json, RouteOptions.class);
   }
 
   /**

--- a/services-directions/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java
+++ b/services-directions/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java
@@ -190,4 +190,28 @@ public class RouteOptionsTest extends TestUtils {
 
     assertEquals(DirectionsCriteria.OVERVIEW_SIMPLIFIED, route.routeOptions().overview());
   }
+
+  @Test
+  public void toJson_fromJson() throws Exception {
+    Response<DirectionsResponse> response = MapboxDirections.builder()
+      .geometries(DirectionsCriteria.GEOMETRY_POLYLINE6)
+      .baseUrl(mockUrl.toString())
+      .accessToken(ACCESS_TOKEN)
+      .origin(Point.fromLngLat(1.0, 1.0))
+      .destination(Point.fromLngLat(5.0, 5.0))
+      .profile(DirectionsCriteria.PROFILE_WALKING)
+      .continueStraight(false)
+      .language(Locale.CANADA)
+      .overview(DirectionsCriteria.OVERVIEW_SIMPLIFIED)
+      .alternatives(true)
+      .build()
+      .executeCall();
+
+    RouteOptions routeOptions = response.body().routes().get(0).routeOptions();
+
+    String jsonString = routeOptions.toJson();
+    RouteOptions routeOptionsFromJson = RouteOptions.fromJson(jsonString);
+
+    assertEquals(routeOptions, routeOptionsFromJson);
+  }
 }

